src/integrationTest/java/com/mongodb/hibernate/BasicInsertionTests.java

```diff
@@ -0,0 +1,176 @@
+import jakarta.persistence.Entity;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Consumer;
+import org.bson.BsonDocument;
+import org.hibernate.cfg.Configuration;
+import org.hibernate.testing.orm.junit.DomainModel;
+import org.hibernate.testing.orm.junit.SessionFactory;
+import org.hibernate.testing.orm.junit.SessionFactoryScope;
+import org.jspecify.annotations.NullUnmarked;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+@NullUnmarked
+@SessionFactory(exportSchema = false)
+@DomainModel(annotatedClasses = {BasicInsertionTests.Book.class, BasicInsertionTests.BookWithEmbeddedField.class})
+class BasicInsertionTests {
+
+    @BeforeEach
+    void setUp() {
+        onMongoCollection(MongoCollection::drop);
+    }
+
+    @Test
+    void testSimpleEntityInsertion(SessionFactoryScope scope) {
+        scope.inTransaction(session -> {
+            var book = new Book();
+            book.id = 1;
+            book.title = "War and Peace";
+            book.author = "Leo Tolstoy";
+            book.publishYear = 1867;
+            session.persist(book);
+        });
+        var expectedDocument = BsonDocument.parse(
+                """
+                        {
+                            _id: 1,
+                            title: "War and Peace",
+                            author: "Leo Tolstoy",
+                            publishYear: 1867
+                        }""");
+        assertCollectionContainsOnly(expectedDocument);
+    }
+
+    @Test
+    void testEntityWithNullFieldValueInsertion(SessionFactoryScope scope) {
+        scope.inTransaction(session -> {
+            var book = new Book();
+            book.id = 1;
+            book.title = "War and Peace";
+            book.publishYear = 1867;
+            session.persist(book);
+        });
+        var expectedDocument = BsonDocument.parse(
+                """
+                        {
+                            _id: 1,
+                            title: "War and Peace",
+                            author: null,
+                            publishYear: 1867
+                        }""");
+        assertCollectionContainsOnly(expectedDocument);
+    }
+
+    @Test
+    void testEntityWithEmbeddedFieldInsertion(SessionFactoryScope scope) {
+        scope.inTransaction(session -> {
+            var book = new BookWithEmbeddedField();
+            book.id = 1;
+            book.title = "War and Peace";
+            book.author = new Author("Leo", "Tolstoy");
+            book.publishYear = 1867;
+            session.persist(book);
+        });
+        var expectedDocument = BsonDocument.parse(
+                """
+                        {
+                            _id: 1,
+                            title: "War and Peace",
+                            authorFirstName: "Leo",
+                            authorLastName: "Tolstoy",
```

Comment:
@stIncMale `TODO` This is a thread about supporting flattening embedding. Something we should discuss over Zoom.

I did not find any hint in the `Statement`s and `SqlAstNode`s fed to `MqlTranslator`, pointing out to flattening embedding. I wonder if there are customization points in Hibernate ORM that would have allowed us to plug into the process of building the internal representation of the domain model (I don't know how Hibernate ORM calls it), where we could react to `@Embeddable`.

Reply:
As we discussed, `@Embeddable` alone has exact meaning just as the testing case shows. I am 100% sure about this. Only very recently it is used for the complex UDT or Struct (it is also used for Array), but that is special.

You can refer to its Javadoc or JPA spec, there is only one meaning of `@Embeddable` (used alone).

Embedded ad-hoc documents or array into one single field is advanced db feature not all SQL vendors supports. As Gavin Kiing mentioned, it is for HQL tech reason that `@Embeddable` is reused for such case (the focus on milestone 2), but there is no confusion when `@Embeddable` is used without `@Struct(type = '...')` used in combination.

Reply:
`@Embeddable` is a simple and basic JPA feature as an abstraction on top of SQL transation. So in this lowest level, Hibernate has finished flattening the fields beforehand. The semantics of `@Embeddable` is abstracted far above such lower-level details. I don't think we need to customize any more.

Reply:
as we discussed at `hibernate` slack channel, we could work around Hibernate's default implementation in m2. At least we could throw exception but maybe we could do better than that. Let us see.